examples/less_than.rs

```diff
@@ -0,0 +1,234 @@
+use bellpepper_core::{
+  boolean::AllocatedBit, num::AllocatedNum, Circuit, ConstraintSystem, LinearCombination,
+  SynthesisError,
+};
+use ff::{PrimeField, PrimeFieldBits};
+use pasta_curves::Fq;
+use spartan2::{
+  errors::SpartanError,
+  traits::{snark::RelaxedR1CSSNARKTrait, Group},
+  SNARK,
+};
+
+fn num_to_bits_le_bounded<F: PrimeField + PrimeFieldBits, CS: ConstraintSystem<F>>(
+  cs: &mut CS,
+  n: AllocatedNum<F>,
+  num_bits: u8,
+) -> Result<Vec<AllocatedBit>, SynthesisError> {
+  let opt_bits = match n.get_value() {
+    Some(v) => v
+      .to_le_bits()
+      .into_iter()
+      .take(num_bits as usize)
+      .map(Some)
+      .collect::<Vec<Option<bool>>>(),
+    None => vec![None; num_bits as usize],
+  };
+
+  // Add one witness per input bit in little-endian bit order
+  let bits_circuit = opt_bits.into_iter()
+    .enumerate()
+    // AllocateBit enforces the value to be 0 or 1 at the constraint level
+    .map(|(i, b)| AllocatedBit::alloc(cs.namespace(|| format!("b_{}", i)), b).unwrap())
+    .collect::<Vec<AllocatedBit>>();
+
+  let mut weighted_sum_lc = LinearCombination::zero();
+  let mut pow2 = F::ONE;
+
+  for bit in bits_circuit.iter() {
+    weighted_sum_lc = weighted_sum_lc + (pow2, bit.get_variable());
+    pow2 = pow2.double();
+  }
+
+  cs.enforce(
+    || "bit decomposition check",
+    |lc| lc + &weighted_sum_lc,
+    |lc| lc + CS::one(),
+    |lc| lc + n.get_variable(),
+  );
+
+  Ok(bits_circuit)
+}
+
+fn get_msb_index<F: PrimeField + PrimeFieldBits>(n: F) -> u8 {
+  n.to_le_bits()
+    .into_iter()
+    .enumerate()
+    .rev()
+    .find(|(_, b)| *b)
+    .unwrap()
+    .0 as u8
+}
+
+// Range check: constrains input < `bound`. The bound must fit into
+// `num_bits` bits (this is asserted in the circuit constructor).
+// Important: it must be checked elsewhere (in an overarching circuit) that the
+// input fits into `num_bits` bits - this is NOT constrained by this circuit
+// in order to avoid duplications (hence "unsafe"). Cf. LessThanCircuitSafe for
+// a safe version.
+#[derive(Clone, Debug)]
+struct LessThanCircuitUnsafe<F: PrimeField> {
+  bound: F, // Will be a constant in the constraits, not a variable
+  input: F, // Will be an input/output variable
+  num_bits: u8,
+}
+
+impl<F: PrimeField + PrimeFieldBits> LessThanCircuitUnsafe<F> {
+  fn new(bound: F, input: F, num_bits: u8) -> Self {
+    assert!(get_msb_index(bound) < num_bits);
+    Self {
+      bound,
+      input,
+      num_bits,
+    }
+  }
+}
+
+impl<F: PrimeField + PrimeFieldBits> Circuit<F> for LessThanCircuitUnsafe<F> {
+  fn synthesize<CS: ConstraintSystem<F>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+    assert!(F::NUM_BITS > self.num_bits as u32 + 1);
+
+    let input = AllocatedNum::alloc(cs.namespace(|| "input"), || Ok(self.input))?;
+
+    let shifted_diff = AllocatedNum::alloc(cs.namespace(|| "shifted_diff"), || {
+      Ok(self.input + F::from(1 << self.num_bits) - self.bound)
```

Comment:
Then this would change to:
```suggestion
      Ok(self.input + F::from(1 << self.num_bits - self.bound))
```